Fix nil upcase on startup

gfiorav opened the pull request:
## Context

A debug trace present in the HTTP header, asked for `ActionDispatch::Http::Headers#[nil]`. This results in an error trace:

```
NoMethodError (undefined method `upcase' for nil:NilClass):
  lib/carto/http_header_authentication.rb:81:in `header_value'
  lib/carto/http_header_authentication.rb:9:in `valid?'
  app/controllers/application_controller.rb:363:in `http_header_authentication?'
```

This PR provides a safety check where `ActionDispatch::Http::Headers#[possible_nil]` is only attempted if `possible_nil` is `.present?`, which means that it's neither `nil` (null) or `''` (empty string).

If the trace is still useful and must be kept (it doesn't provide much info), this check can be used to not generate the error when rendering it.

This PR also provides test to check that the code is indeed nil-resistant.

gfiorav committed:
make header_value nil-header resistant

gfiorav committed:
add test to check load with nil header won't break

gfiorav committed:
add optional param to stub header values

vishalya commented on lib/carto/http_header_authentication.rb:
```diff
@@ -85,8 +85,7 @@ def field_from_value(request)
       header = ::Cartodb.get_config(:http_header_authentication, 'header')
-      puts "user-auto-creation : Trying to extract value from headers for #{header}, value is #{headers[header]}"
-      !header.nil? && !header.empty? ? headers[header] : nil
+      headers[header] if header.present?
```
We need the logs for the 'user-auto-creation', as it traces the request for the user authentication and auto-creation. the idea is that, if you search the logs for 'user-auto-creation, it should give the flow of the request - if something fails. You can either put this logging back or come up with a better solution. I agree, it looks ugly, but it serves a useful purpose.

gfiorav replied:
We do this by logging NGINX requests. We then query them with Elastic Search / Kibana.

Please be aware that you shouldn't trust these flows, since when multithreading, some messages might not make it to stdout.

gfiorav replied:
In the mean time, I'm putting the log back in. Do you need the values or were those for debugging? It be simpler to put `user-auto-creation : Trying to extract value from headers`, no interpolation needed.

What do you think?

vishalya replied:
yes, that should be good.

vishalya commented on spec/lib/carto/http_header_authentication_spec.rb:
```diff
@@ -43,6 +43,12 @@
       Carto::HttpHeaderAuthentication.new.valid?(mock_email_request).should == true
     end
+
+    it 'resists when config header is nil' do
```
What's the command you run to test it? 
Also, are you able to kick off the auto-testing inside the VM? Please give me the command if you are able to do it?

gfiorav replied:
Sure!

RAILS runs tests in separate environment, `test`, that has to be migrated like the `development` or `production` environments. We provide a scrip in the `Makefile` for running all tests:

```
make check
```

This will create your testing DBs and migrate the to the latest schema before running all specs.

Manually, you'd do:
```
RAILS_ENV=test bundle exec rails db:drop
RAILS_ENV=test bundle exec rails db:create
RAILS_ENV=test bundle exec rails db:migrate
```

Now your testing environment is ready. Please be aware of the `RAILS_ENV=test`, if you don't specify it, `RAILS_ENV=development` is the default and it will wipe your db in `development`.

Now, to run any single spec (this one for exapmple), simply:
```
bundle exec rspec spec/lib/carto/http_header_authentication_spec.rb -e "resists when config header is nil"
```

Note that we don't specify the environment since the `rspec` gem sets `test` by default. You can also specify the line number where the test is at instead of its name.

gfiorav committed:
add logging back in

gfiorav commented:
Hey @vishalya, I added the logging back in.

See if it's good enough :)

vishalya commented:
I would like to see the values `for #{header}, value is #{headers[header]}"` if they are available... else simply hardcode to `Not available`

gfiorav committed:
de-proxy submodules

gfiorav commented:
Ok. Btw, what do you think about the multithreading issue?

Also, we use this for logging: lib/carto/http_header_authentication.rb

In production it works with Rollbar but locally it prints to std out.

Question: what do you parse to get these messages? `log/<environment>.log`?

vishalya commented:
Regarding multithreading, it's not complete. Logging is work in progress.
The way we work with logs in production is that - we collect all the logs from all the servers (nginx, rails, windshaft/sql api etc.) to kibana logstash and then we search for a key... in this auto-user-creation it will be the username and see if there are any issues.

gfiorav committed:
go back to previous log

gfiorav commented:
Ok @vishalya, I think it's how you want it now

gfiorav committed:
Better comment

vishalya commented:
Looks good, thanks, I will merge it after a round of integration testing, for the next sprint.

gfiorav committed:
Merge remote-tracking branch 'guido/fix-nil-upcase-on-startup' into HEAD

coyle5280 committed:
alias updates table ui

coyle5280 committed:
alias updates add backend rb code, migration and tests

coyle5280 committed:
add alias name to filters and templates

coyle5280 committed:
add alias information to infowindows

coyle5280 committed:
add alias info to table header and table title
update wizard.js and widgets.js

coyle5280 committed:
add alias info to tooltips

coyle5280 committed:
point cdb module to bloomberg_features branch

gfiorav committed:
bring extra changes in backend

gfiorav committed:
use single json aliases column

coyle5280 committed:
Modify Table Alias for new backend obj

coyle5280 committed:
update alias columns for new aliases json

coyle5280 committed:
removed filter toggle code

coyle5280 committed:
delay column table build until table alias data loaded

coyle5280 committed:
rename carto_db id column to ref_id

coyle5280 committed:
enlarge table column header

coyle5280 committed:
only show alias name if read_only table

coyle5280 committed:
lint and code cleanup

coyle5280 committed:
Go to Map Library bug fix

gfiorav committed:
split aliases into name_alias and column_aliases for better search performance

coyle5280 committed:
Seperate table alias and column aliases

coyle5280 committed:
Merge branch 'aliases' of https://github.com/cartodb-org/cartodb into aliases

coyle5280 committed:
'Update Aliases for two columns'

Commit:
uncommit .gitmodules

gfiorav committed:
further remove .gitmodules modifications

mbektas committed:
Data & Map Library changes, side menu, category tree, initial commit

coyle5280 committed:
Column updates for development cdb module

coyle5280 committed:
Add in loaded notice

coyle5280 committed:
Merge branch 'aliases' of https://github.com/cartodb-org/cartodb into aliases

gfiorav committed:
styling

gfiorav committed:
styling on tabledata.js

gfiorav committed:
wizard.js cleanup

gfiorav committed:
little cleanups here and there

gfiorav committed:
style + refactor

gfiorav committed:
further cleanup

gfiorav committed:
remove extra comment

mbektas committed:
fix for disappearing filters panel on Map to Table switch

gfiorav committed:
bring ghost tables manager + bolt

mbektas committed:
declare main parent category ids

gfiorav committed:
bring bolt and ghost tables manager specs

gfiorav committed:
add specs to Makefile

gfiorav committed:
uncommit .gitmodules

gfiorav committed:
Merge remote-tracking branch 'bbg/blp_dev' into ghost_table_manager

gfiorav committed:
sync with blp_dev

gfiorav committed:
sync with blp_dev

gfiorav committed:
remove header auth

vmanohar1 committed:
Merge pull request #139 from cartodb-org/ghost_table_manager

Ghost table manager + Bolt

gfiorav committed:
fix typo renameAlais -> renameAlias

gfiorav committed:
revert lib/carto/http_header_authentication.rb

gfiorav committed:
revert spec/lib/carto/http_header_authentication_spec.rb

gfiorav committed:
revert spec/requests/http_authentication_helper.rb

mbektas committed:
Merge pull request #138 from cartodb-org/aliases

Alias feature

mbektas committed:
update submodule cartodb.js

coyle5280 committed:
remove column type from filter drop down

coyle5280 committed:
Only  users with aliases feature flag can add aliases

coyle5280 committed:
Merge branch 'aliases' of https://github.com/cartodb-org/cartodb into aliases

coyle5280 committed:
fixed bug close header dropdown on title click

coyle5280 committed:
http_header_fix

gfiorav committed:
remove commented code

gfiorav committed:
remove extra spaces

gfiorav committed:
indentation fixes

gfiorav committed:
use built-in user#featureEnabled method for ff check

gfiorav committed:
use user#featureEnabled instead of helper

coyle5280 committed:
Added alias name to menu layer

coyle5280 committed:
maintain alias layer name when switching between map and table

gfiorav committed:
use user#featureEnabled for ff check

gfiorav committed:
use user passed to template

gfiorav committed:
use template var

coyle5280 committed:
corrected aliases feature flag implementation

coyle5280 committed:
remove tooltips from dataset views

coyle5280 committed:
Fix Sugar Not Found Filter List Error

coyle5280 committed:
code cleanup

coyle5280 committed:
removed Title from a tag to prevent tooltip

coyle5280 committed:
Undo changes to datasets_item.js

gfiorav committed:
Merge pull request #137 from cartodb-org/135_Sugar_bug_fix

135 sugar bug fix

gfiorav committed:
Merge pull request #141 from cartodb-org/134_remove_tooltip_datasets

134 remove tooltip datasets

gfiorav committed:
remove commented code

gfiorav committed:
revert cosmetic change

mbektas committed:
Merge pull request #140 from cartodb-org/aliases

Aliases Feedback fixes

mbektas committed:
data library toolbar changes

coyle5280 committed:
remove ref to table_name_aliases and replace

coyle5280 committed:
uncommit files

coyle5280 committed:
remove http_header file

coyle5280 committed:
Only data users in edit mode can add or edit aliases

coyle5280 committed:
roll back cdb to commit 52b4941

gfiorav committed:
remove extra line

gfiorav committed:
remove extra spaces

gfiorav committed:
Merge pull request #143 from cartodb-org/142_Show_aliases_only_from_table_view

Only data users in edit mode can add or edit aliases

gfiorav committed:
Merge pull request #136 from cartodb-org/133_add_alias_to_layer_panel

133 add alias to layer panel

gfiorav committed:
Merge branch 'source' into aliases

gfiorav committed:
Revert "Merge branch 'source' into aliases"

This reverts commit d6a4f1aeaba55dcafee488c7fcbccef618f3bfad, reversing
changes made to c0d66bacd02beb5c5d7c5c981b362bd213299c9e.

mbektas committed:
dataset library - page layout changes

mbektas committed:
Merge pull request #141 from cartodb-org/aliases

Aliases feedback fixes (part II)

mbektas committed:
layout changes, remove onboarding

mbektas committed:
fix for menu hilite states

mbektas committed:
updated side menu class names

mbektas committed:
rake tasts for dataset alias sync

klibloomberg committed:
updated legend background style to match gp legends better

mbektas committed:
Merge pull request #143 from bloomberg/legendstyleupdate3.13

updated legend background style to match gp legends better

mbektas committed:
fixes for toolbar button states

mbektas committed:
db migrate scripts for categories

mbektas committed:
rake tasks for category sync

mbektas committed:
Revert "rake tasks for category sync"

This reverts commit 8e47e122d7521b99593dd1555389ff88d58c9cb4.

mbektas committed:
Merge pull request #144 from bloomberg/new_data_library

Merge new dataset & map library into blp_dev

mbektas committed:
rake tasks for category sync

gfiorav committed:
expose name alias in visulization#index

coyle5280 committed:
column alias as null when undefined

coyle5280 committed:
added table alias name to datasets and template files

gfiorav committed:
decide title in model

gfiorav committed:
don't decide on template

gfiorav committed:
fix ff name

gfiorav committed:
remove feature flag checks

gfiorav committed:
add alias to old presenter

mbektas committed:
fix category initialization problems

gfiorav committed:
Merge branch 'source' into 152_aliased_tablename_appear_in_dashboard

mbektas committed:
dont hide toolbar and side menu when creating new map

gfiorav committed:
load name alias into visualization display name for common data

gfiorav committed:
Merge branch 'source' into 152_aliased_tablename_appear_in_dashboard

gfiorav committed:
remove table fetch and add explanation

gfiorav committed:
remove user load

gfiorav committed:
remove comment

gfiorav committed:
add extra line back

dprophet committed:
patch importer and geocoder

gfiorav committed:
Merge pull request #154 from cartodb-org/152_aliased_tablename_appear_in_dashboard

added table alias name to datasets and template files

gfiorav committed:
Merge remote-tracking branch 'bloomberg/blp_dev' into source

mbektas committed:
Merge pull request #146 from cartodb-org/source

Show Aliased dataset names in Datasets dashboard

mbektas committed:
watch css files in cartodb.js and fix problem with copy:vendor

mbektas committed:
rake task to set category by name

coyle5280 committed:
update alias config when column name changes

coyle5280 committed:
update template for proper column renaming logic

coyle5280 committed:
update filter dropdown to handle null alias value

coyle5280 committed:
update header view js to handle new rename logic

gfiorav commented:
@vishalya since I see conflicts have appeared... is this still on course to being merged?

vishalya commented:
I think, Mehmet might have merged it, if it was a part of any other PR, Let me check with him and get back to you.

mbektas committed:
show underline for dataset title when editable, only edit table types

mbektas committed:
prevent setting map category inadvertently

mbektas committed:
remove unnecessary SECURITY DEFINER

gfiorav committed:
Show original name for mapdata user

This commit adds a method _title that handles the logic so that the
original name will be shown in parenthesis for users with the 'aliases'
feature flag enabled. So:

Normal user:
  Alias

Mapsdata user:
  Alias (original_dataset_name)

gfiorav committed:
add missing +

vishalya committed:
Merge pull request #145 from bloomberg/importer_geocoder

patch importer and geocoder

gfiorav committed:
Remove extra white line

coyle5280 committed:
update var org_col_name to original_col_name

coyle5280 committed:
update template to reflect new original_col_name var

mbektas committed:
Merge pull request #147 from cartodb-org/show_original_name_for_mapsdata_user

Show original name for mapsdata user

coyle5280 committed:
replace ternary with coalesce operation

gfiorav committed:
Merge branch 'source' into 149_column_rename

coyle5280 committed:
update common dataset list item js files to show table alias

coyle5280 committed:
Update common jst.ejs datalist templates to show table alias

coyle5280 committed:
change empty alias value to null

coyle5280 committed:
header dropdown now looks for null alias value to display add

mbektas committed:
fixes for new layout UI issues, temporary search functionality

gfiorav committed:
Default to display_name

Grabbing display_name was previous behaviour, so we default to it to
avoid null issues later on. It would be best to make sure that those
pieces that break with null parmas are well protected.

gfiorav committed:
a bit of formatting

gfiorav committed:
Merge branch 'source' into cb-233-default-to-dataset-name

mbektas committed:
Merge pull request #148 from cartodb-org/cb-233-default-to-dataset-name

[ALIASES] Default to dataset name

gfiorav committed:
remove second line from dataset name tiles

gfiorav committed:
Merge branch 'source' into 156_alias_names_browse_modal

gfiorav committed:
preform name with aliases

gfiorav committed:
fixed stacking

mbektas committed:
Merge pull request #149 from cartodb-org/149_column_rename

[ALIASES] Rename column fix

gfiorav committed:
remove duplicate title

gfiorav committed:
fix duplicate call

gfiorav committed:
fix typo, remove new

gfiorav committed:
check if change will be redundant

mbektas committed:
handle categories without sub-categories

mbektas committed:
remove dataset item tooltips

mbektas committed:
change order of Upload Dataset and Open in Map, rename toolbar and menu items

gfiorav committed:
Merge pull request #155 from cartodb-org/149_column_rename

149 column rename

gfiorav committed:
Merge pull request #157 from cartodb-org/156_alias_names_browse_modal

156 alias names browse modal

mbektas committed:
Merge pull request #150 from cartodb-org/156_alias_names_browse_modal

[ALIASES] Show aliases in browse modal

mbektas committed:
remove quotes

gfiorav committed:
Merge remote-tracking branch 'bloomberg/blp_dev' into source

mbektas committed:
make description column wider

coyle5280 committed:
add alias feature flag check for original table name

mbektas committed:
Merge pull request #151 from cartodb-org/158_hide_orig_tablename

[ALIASES] Hide original name for non-mapsdata users

gfiorav committed:
remove nil check

mbektas committed:
Merge pull request #152 from cartodb-org/fix-category-selector

[ALIASES] Fix categories dropdown

mbektas committed:
make loading animation area shorter

mbektas committed:
fixes for loading error and no dataset messages

mbektas committed:
enable Open in Map also for favorites, show locked maps in All

gfiorav commented:
Can we please merge this into blp_dev 🙏 !

gfiorav committed:
Merge branch 'blp_dev' into fix-nil-upcase-on-startup

gfiorav committed:
Merge branch 'blp_dev' into fix-nil-upcase-on-startup

gfiorav commented:
Closing due to inactivity
